Add a new fusion instruction and remove unnecessary ones

qwe661234 opened the pull request:
After constant propagation and constant folding, we can optimize some ALU or brnach instructions into simple instructions.

Close: #226

jserv commented on src/emulate.c:
```diff
@@ -834,102 +823,102 @@ static bool detect_memcpy(riscv_t *rv, int lib)
     return true;
 }
 
+static bool stdlib_injection(riscv_t *rv, block_t *block)
```
Rename to `libc_substitute`.

jserv reviewed (changes requested):
Consistently employ the prefix `constopt_` rather than `const_opt` to prevent potential confusion.

jserv commented on src/const_opt.c:
```diff
@@ -0,0 +1,1009 @@
+/* RV32I Base Instruction Set */
```
Rename to `src/rv32_constopt.c`.

jserv commented on src/emulate.c:
```diff
@@ -957,25 +961,46 @@ static void match_pattern(riscv_t *rv, block_t *block)
+    uint32_t const_val[32];
+} const_opt_info_t;
+
+#define CONSTOPT(inst, code)                                              \
```
Do `#undef CONSTOPT` when the code generation completes.

jserv commented:
Both constant folding and constant propagation are related compiler optimization techniques. Constant propagation replaces the bound variable with the constant expression it is bound to. Conversely, constant folding evaluates the expression with all compile-time inputs. I am a bit confused about the proposed changes, as I cannot ensure that a fully functional implementation exists for either constant folding or constant propagation. Can you verify this?

jserv commented on src/emulate.c:
```diff
@@ -957,25 +961,46 @@ static void match_pattern(riscv_t *rv, block_t *block)
+#undef _
+};
+/* clang-format on */
+typedef void (*opt_func)(rv_insn_t *, const_opt_info_t *);
```
Should be `constopt_func_t`.

jserv commented on src/emulate.c:
```diff
@@ -957,25 +961,46 @@ static void match_pattern(riscv_t *rv, block_t *block)
+};
+/* clang-format on */
+typedef void (*opt_func)(rv_insn_t *, const_opt_info_t *);
+static void constant_opt(block_t *block)
```
Rename to `optimize_constant` for consistency.

jserv commented on src/emulate.c:
```diff
@@ -957,25 +961,46 @@ static void match_pattern(riscv_t *rv, block_t *block)
 
+
+typedef struct const_opt_info {
+    bool is_constant[32];
```
Why hard-coded `32`?

qwe661234 committed:
Add constant propagation and constant folding

With constant propagation and constant folding, we can optimize some ALU
or brnach instructions into simple instructions.

See: #226

jserv commented on src/emulate.c:
```diff
@@ -834,102 +823,102 @@ static bool detect_memcpy(riscv_t *rv, int lib)
+            }
+        }
+        break;
+        /* TODO: inject other frequently-used standard library */
```
Rephrase as following:
> Inject other frequently used function calls from the C standard library.

jserv commented on src/emulate.c:
```diff
@@ -957,25 +961,48 @@ static void match_pattern(riscv_t *rv, block_t *block)
 }
 
+
+typedef struct constopt_info {
```
Drop the use of `constopt_info`. i.e., use anonymous structure instead.

qwe661234 commented:
> Both constant folding and constant propagation are related compiler optimization techniques. Constant propagation replaces the bound variable with the constant expression it is bound to. Conversely, constant folding evaluates the expression with all compile-time inputs. I am a bit confused about the proposed changes, as I cannot ensure that a fully functional implementation exists for either constant folding or constant propagation. Can you verify this?

There is a simple example in `./build/dhrystone.elf` can be optimized by constant folding and constant propagation.
* Before
```
   0x100f0:	auipc	gp,0x18
   0x100f4:	add	gp,gp,864 
   0x100f8:	add	a0,gp,-1996
   0x100fc:	auipc	a2,0x1a
   0x10100:	add	a2,a2,1384 #
   0x10104:	sub	a2,a2,a0
```
* After
```
   0x100f0:	lui	gp,  0x10108        
   0x100f4:	lui	gp,  0x10468
   0x100f8:	lui	a0, 0xFC9C
   0x100fc:	lui	a2,0x10116
   0x10100:	lui	a2, 0x1067E 
   0x10104:	lui    a2, 0x9e2
```
Moreover, these 6 `lui` instructions can be fused into a fused instruction.

jserv commented:
> There is a simple example in `build/dhrystone.elf` can be optimized by constant folding and constant propagation.
[...]
> Moreover, these 6 `lui` instructions can be fused into a fused instruction.

Describe this within git commit messages.

jserv commented on src/emulate.c:
```diff
@@ -957,25 +962,48 @@ static void match_pattern(riscv_t *rv, block_t *block)
 
+
+typedef struct {
+    bool is_constant[N_RV_REGS];
```
Does it make sense to check x0, which is always zero?

qwe661234 replied:
Added

jserv commented on src/emulate.c:
```diff
@@ -957,25 +962,48 @@ static void match_pattern(riscv_t *rv, block_t *block)
+static void optimize_constant(block_t *block)
+{
+    constopt_info_t constopt_info;
+    memset(&constopt_info, 0, sizeof(constopt_info));
```
The C standards guarantee that even if an integer array is located on automatic storage and if there are fewer initializers in a brace-enclosed list then the uninitialized elements must be initialized to 0.

C99 Standard 6.7.8.21
> If there are fewer initializers in a brace-enclosed list than there are elements or members of an aggregate, or fewer characters in a string literal used to initialize an array of known size than there are elements in the array, the remainder of the aggregate shall be initialized implicitly the same as objects that have static storage duration.

You can safely rewrite as following without the invocation of `memset`.
```c
    constopt_info_t info = {0};
```

qwe661234 committed:
Add a new fusion instruction and remove unnecessary ones

After constant propagation and constant folding, auipc + addi and lui +
addi become 2 lui instuctions. Moreover, we discover a sequence of lui
instrution existing in optimized basic block, so we add a new fusion
function to handle multiple lui instructions.

There is a simple example in ./build/dhrystone.elf can be optimized by
constant folding and constant propagation.

Before
   0x100f0: auipc gp,0x18
   0x100f4: add gp,gp,864
   0x100f8: add a0,gp,-1996
   0x100fc: auipc a2,0x1a
   0x10100: add a2,a2,1384
   0x10104: sub a2,a2,a0
After
   0x100f0: lui gp, 0x10108
   0x100f4: lui gp, 0x10468
   0x100f8: lui a0, 0xFC9C
   0x100fc: lui a2,0x10116
   0x10100: lui a2, 0x1067E
   0x10104: lui a2, 0x9e2

Moreover, these 6 lui instructions can be fused into a fused instruction.
